### PR TITLE
Add rel attribute for Social Icons

### LINKED ISSUE #38630: Allow rel customizations in social link block
Currently, the social link block has some hard-coded rel attributes that are applied only if the parent social links block is set to open in a new tab. When the open in new tab toggle is set to true, each link receives `rel="noopener nofollow"` if it is not selected there is no rel attribute. For example, it's not a guarantee that if a link will open in a new tab that it should be marked nofollow and that if it is not set to open in a new tab it shouldn't have a nofollow rel. The use case would be a theme block with template parts loading social links in a theme, the user may want to edit the rel values for SEO or other purposes. See the implementation here: https://github.com/WordPress/gutenberg/blob/55ca64fec89a58d370675a9dba6a7914c8920e49/packages/block-library/src/social-link/index.php#L33

Ideally, the social link block can have a custom rel attribute. This is a pattern established in the button block already, and users will expect consistency in how link patterns are treated and managed. The social link should apply smart rel attributes automatically but also allow the user to customize the attribute. Placing the rel field within the advanced block settings is a great pattern to follow established by the button block.

Existing implementation of established pattern in button block:
https://github.com/WordPress/gutenberg/blob/55ca64fec89a58d370675a9dba6a7914c8920e49/packages/block-library/src/button/edit.js#L33
https://github.com/WordPress/gutenberg/blob/55ca64fec89a58d370675a9dba6a7914c8920e49/packages/block-library/src/button/edit.js#L86-L107
https://github.com/WordPress/gutenberg/blob/55ca64fec89a58d370675a9dba6a7914c8920e49/packages/block-library/src/button/edit.js#L256-L260

### PR DESCRIPTION
**Pull request by georgeh**
## What?
Makes `rel` attribute editable on social icons, closes #38630

## Why?

Some social networks like IndieWeb and Mastodon use the pattern of looking for a link back with `rel="me"` to establish that a user controls a URL. Adding the option for a custom `rel` attribute allows people to claim their WordPress sites using Gutenberg.

## How?
This mimics the implementation of the `rel` attribute on the Button block, by exposing the attribute as an advanced option in the block inspector.

## Testing Instructions
1. Add a Social Links block to a page or template
2. Choose the block and enter something for the `rel`
3. Save
4. Inspect the element on the rendered page and confirm that it has the correct `rel` attribute on the `<a>` element

## Screenshots or screencast 
Screenshot 2022-11-01 at 3 27 17 PM

### CODE REVIEW

File: packages/block-library/src/social-link/index.php [hunk 1]
```diff
@@ -57,13 +53,21 @@ function render_block_core_social_link( $attributes, $content, $block ) {
...
-	return $link;
+	$w = new WP_HTML_Tag_Processor($link);
+	$w->next_tag('a');
+	if ( $open_in_new_tab ) {
+		$w->set_attribute('rel', esc_attr( $rel ) . ' noopener nofollow');
+		$w->set_attribute('target', '_blank');
+	} else if ( $rel !== '' ) {
+		$w->set_attribute('rel', esc_attr( $rel) );
+	}
+	return $w;
```
Comment (dmsnell) [inline packages/block-library/src/social-link/index.php:70]: nice use of the tag processor!

**Review by dmsnell — APPROVED** [review]
I see no problem with this, unless someone finds the "Link rel" label confusing. If that's the case it's easy to improve it as we go.

File: packages/block-library/src/social-link/index.php (hunk 1 above)
Comment (georgeh) [reply packages/block-library/src/social-link/index.php:70]: It seemed like a clearer solution than a bunch of string manipulation and worked like a charm!

File: packages/block-library/src/social-link/edit.js [hunk 2]
```diff
@@ -88,6 +88,13 @@ const SocialLinkEdit = ( {
 		},
 	} );
 
+	const onSetLinkRel = useCallback(
+		( value ) => {
+			setAttributes( { rel: value } );
+		},
+		[ setAttributes ]
+	);
```
Comment (Mamaduka) [inline packages/block-library/src/social-link/edit.js]: Suggestion (Non-blocking): This is probably unnecessary optimization. We're not passing callback deep down the tree, and `TextControl` will re-render anytime `SocialLinkEdit` re-renders.

File: packages/block-library/src/social-link/index.php (hunk 1 above)
Comment (Mamaduka) [reply packages/block-library/src/social-link/index.php:70]: Nit: While I also like how clean the "tag processor" looks here, but I don't think it is needed. We're building whole block output in the render callback; in cases like this, I would prefer PHP as a templating engine.

**Review by Mamaduka — APPROVED** [review]
This looks great, @georgeh!

I left two suggestions, but feel free to ignore them.

**Comment by Mamaduka** [comment]
@dmsnell, the 'Link rel' will make this consistent with other blocks.

There's also a suggestion to add a description to the link relationship settings. See https://github.com/WordPress/gutenberg/issues/20208#issuecomment-790928125.

File: packages/block-library/src/social-link/edit.js (hunk 2 above)
Comment (georgeh) [reply packages/block-library/src/social-link/edit.js]: Good catch, cleaned that up.

File: packages/block-library/src/social-link/index.php (hunk 1 above)
Comment (dmsnell) [reply packages/block-library/src/social-link/index.php:70]: @Mamaduka I'm confused on what you are saying here. I guess I can see that we're constructing the link HTML entirely, we're not modifying content that was given to us, in which case we could change this whole thing to something like…

```php
$anchor = $open_in_new_tab
	? '<a href="' . esc_url( $url ) . '" rel="' . esc_attr( $rel ) . ' noopener nofollow" target="_blank" class="wp-block-social-link">'
	: '<a href="' . esc_url( $url ) . '" class="wp-block-social-link">`;

…

$link .= $anchor;
```

that makes sense, though it does highlight a use-case for the tag processor that maybe wasn't the primary goal when creating it, which is making tag text construction more obvious. personally I often find these template pieces noisy and hard to follow, especially since they tend to make for very long lines with intertwining PHP and strings.

```php
$anchor = new WP_HTML_Tag_Processor( "<a>" );
$anchor->next_tag( 'a' );
$anchor->set_attribute( 'href', $url );
$anchor->set_attribute( 'class', 'wp-block-social-link-anchor' );
if ( $open_in_new_tab ) {
	$anchor->set_attribute( 'rel', "{$rel} no opener nofollow" );
	$anchor->set_attribute( 'target', '_blank' );
} else {
	$anchor->set_attribute( 'rel', $rel );
}
$link .= $anchor;
```

now I think maybe there could be a good helper for this 🤔 

```php
static function make_tag_opener( $tag_name, $attributes = [] ) {
	$w = new WP_HTML_Tag_Walker( "<{$tag_name}>" );
	$w->next_tag( $tag_name );

	foreach ( $attributes as $name => $value ) {
		$w->set_attribute( $name, $value );
	}

	return $w;
} 
```

this code here would then turn into something even less noisy, and at the same time, properly encoded.

```php
$anchor = WP_HTML_Tag_Processor::make_tag_opener( 'a', array(
	'href'  => $url,
	'class' => 'wp-block-social-anchor',
	'rel'   => $rel
) );

if ( $open_in_new_tab ) {
	$anchor->set_attribute( 'rel', "{$rel} no opener nofollow" );
	$anchor->set_attribute( 'target', '_blank' );
}

…

$link .= $anchor;
```

none of this is to contradict your input; just taking the opportunity to muse on how this could evolve. this is a very DOM-like interface with its benefits as well as its drawbacks.

cc: @adamziel

Comment (Mamaduka) [reply packages/block-library/src/social-link/index.php:70]: @dmsnell, sorry about the confusion.

> I guess I can see that we're constructing the link HTML entirely, we're not modifying content that was given to us, in which case we could change this whole thing to something like

This is what I was trying to formulate but failed 😅

**Comment by james-s-k** [comment]
When adding "me" or any other text in the rel field, it does not remove "noopener nofollow". So the rel attribute now looks like this: rel="me noopener nofollow". Wouldn't I want Google to follow the link if I am adding "me"?

`
<a rel="me noopener nofollow" target="_blank" href="https://www.facebook.com/xxxxxxx" class="wp-block-social-link-anchor">`
